sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsDeviceMemoryStore.scala

```diff
@@ -256,35 +267,88 @@ class RapidsDeviceMemoryStore(catalog: RapidsBufferCatalog = RapidsBufferCatalog
     addBuffer(buffer)
   }
 
+  /**
+   * The RapidsDeviceMemoryStore is the only store that supports setting a buffer spillable
+   * or not.
+   */
+  override protected def setSpillable(buffer: RapidsBufferBase, spillable: Boolean): Unit = {
+    doSetSpillable(buffer, spillable)
+  }
+
   class RapidsDeviceMemoryBuffer(
       id: RapidsBufferId,
       size: Long,
       meta: TableMeta,
-      table: Option[Table],
       contigBuffer: DeviceMemoryBuffer,
       spillPriority: Long,
       spillCallback: SpillCallback)
-      extends RapidsBufferBase(id, size, meta, spillPriority, spillCallback) {
+      extends RapidsBufferBase(id, size, meta, spillPriority, spillCallback)
+        with MemoryBuffer.EventHandler {
+
     override val storageTier: StorageTier = StorageTier.DEVICE
 
-    override protected def releaseResources(): Unit = {
+    require(contigBuffer.setEventHandler(this) == null,
```

Comment:
If this does throw an exception how bad of a state is the spill code? I mean we just changed who the event handler is pointing to, should we exit? should it be a fatal error? should we not change the event handler unless it is exactly what we expect? I am just a little concerned about how we might recover from this.

Reply:
I have changed the call instantiating `RapidsDeviceMemoryBuffer` to hold the `contigBuffer` lock. Before instantiating this we go through the event handler detection logic as well, so this is the new order of operation:

1. Lock the cuDF memory buffer
2. Try to get a handler from the memory buffer. If not set, we will create a new `RapidsDeviceMemoryBuffer`. If set, we call `addReference` on the `RapidsDeviceMemoryBuffer` and if that succeeds, we know we were valid and are _at worst_ in the pending queue to be spilled. If we can't add a reference we lost the race, and the buffer has since spilled to host, or has been removed.
3. We instantiate a new `RapidsDeviceMemoryBuffer` if (2) failed to find a handler, or failed to `addReference`.
4. We associate this new `RapidsDeviceMemoryBuffer` with the handler.
5. Unlock cuDF memory buffer

I do not think it's possible we wind up here. I'd like this to be a loud exception since it would be a very bad thing if we overwrote that handler. So I expect no recovery from this scenario. @revans2 please let me know if the logic makes sense.